### ISSUE TITLE
program won't compile

### ISSUE THREAD

**Issue by denhamgithub** [sergev/dmrconfig#6]
error I get --
~/md380/dmrconfig $ make
gcc -g -o dmrconfig main.o util.o radio.o dfu-libusb.o uv380.o md380.o rd5r.o gd77.o hid.o hid-libusb.o -Wl,-Bstatic -lusb-1.0 -Wl,-Bdynamic -lpthread -ludev
/usr/bin/ld: cannot find -lusb-1.0
collect2: error: ld returned 1 exit status
Makefile:27: recipe for target 'dmrconfig' failed
make: *** [dmrconfig] Error 1
den@den-475M ~/md380/dmrconfig $

I tried to install some recommendations but they failed. such as apt install libusb-1.0.0-dev libusb-1.0.0
fortunately, apt remove said it removed them. I posted the problem on r/DMR when I did not know how to contact you, Mr. Sergev. But someone on Reddit told me how to do that.

I am on Linux Mint 19 'Tara' based on Ubuntu Bionic 18.04. Kernel is 4.18.4

**Comment by sergev** [comment]
Sorry I missed your post on r/DMR.
The problem here is that gcc cannot find a static library libusb-1.0.a for some reason. A simple workaround still exists: just remove "-Wl,-Bstatic" and "-Wl,-Bdynamic" flags from the LIBS= line in Makefile. In this case a dynamic version (libusb-1.0.so) will be used instead.

**Comment by sergev** [comment]
Sorry, I cannot reproduce your bug on Ubuntu 18.04.1. With libusb package installed, it builds correctly:
```
$ sudo apt install libusb-1.0-0-dev 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'libusb-1.0-0-dev' for regex 'libusb-1.0.0-dev'
The following NEW packages will be installed:
  libusb-1.0-0-dev
0 upgraded, 1 newly installed, 0 to remove and 24 not upgraded.
Need to get 60.5 kB of archives.
After this operation, 285 kB of additional disk space will be used.
Get:1 http://us.archive.ubuntu.com/ubuntu bionic/main amd64 libusb-1.0-0-dev amd64 2:1.0.21-2 [60.5 kB]
Fetched 60.5 kB in 1s (80.9 kB/s)         
Selecting previously unselected package libusb-1.0-0-dev:amd64.
(Reading database ... 347231 files and directories currently installed.)
Preparing to unpack .../libusb-1.0-0-dev_2%3a1.0.21-2_amd64.deb ...
Unpacking libusb-1.0-0-dev:amd64 (2:1.0.21-2) ...
Setting up libusb-1.0-0-dev:amd64 (2:1.0.21-2) ...

$ ls -l /usr/lib/x86_64-linux-gnu/libusb-1.0.a
-rw-r--r-- 1 root root 185176 Jun 24  2017 /usr/lib/x86_64-linux-gnu/libusb-1.0.a

$ make
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o main.o main.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o util.o util.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o radio.o radio.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o dfu-libusb.o dfu-libusb.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o uv380.o uv380.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o md380.o md380.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o rd5r.o rd5r.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o gd77.o gd77.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o hid.o hid.c
gcc -g -O -Wall -Werror -DVERSION='"0.9.133"'   -c -o hid-libusb.o hid-libusb.c
gcc -g -o dmrconfig main.o util.o radio.o dfu-libusb.o uv380.o md380.o rd5r.o gd77.o hid.o hid-libusb.o -Wl,-Bstatic -lusb-1.0 -Wl,-Bdynamic -lpthread -ludev
``` 
Can you please double-check the package libusb-1.0-0-dev is installed on your system? Note: it's "libusb-1.0-0-dev", not "libusb-1.0.0-dev". 
This package contains a static version of the library:
```
$ dpkg -L libusb-1.0-0-dev
...
/usr/lib/x86_64-linux-gnu/libusb-1.0.a
...
```

**Comment by denhamgithub** [comment]
OK, sir.  I'm pretty sure I followed your instructions to the 'letter' so let me show what happened. First, I did not edit the Makefile but will try that next.
-- did the apt install of libusb-1.0-0-dev and got this 
sudo apt install libusb-1.0-0-dev
[sudo] password for den: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  linux-modules-4.15.0-36-generic
Use 'sudo apt autoremove' to remove it.
Recommended packages:
  libusb-1.0-doc
The following NEW packages will be installed:
  libusb-1.0-0-dev
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 60.5 kB of archives.
After this operation, 285 kB of additional disk space will be used.
Get:1 http://la-mirrors.evowise.com/ubuntu bionic/main amd64 libusb-1.0-0-dev amd64 2:1.0.21-2 [60.5 kB]
Fetched 60.5 kB in 0s (350 kB/s)      
Selecting previously unselected package libusb-1.0-0-dev:amd64.
(Reading database ... 441800 files and directories currently installed.)
Preparing to unpack .../libusb-1.0-0-dev_2%3a1.0.21-2_amd64.deb ...
Unpacking libusb-1.0-0-dev:amd64 (2:1.0.21-2) ...
Setting up libusb-1.0-0-dev:amd64 (2:1.0.21-2) ...
(LINE by line, my output, except for 'reading database .... 441800 files) matches what you got)

2nd step -- make still errors with ~/md380/dmrconfig $ make
gcc -g -o dmrconfig main.o util.o radio.o dfu-libusb.o uv380.o md380.o rd5r.o gd77.o hid.o hid-libusb.o -Wl,-Bstatic -lusb-1.0 -Wl,-Bdynamic -lpthread -ludev
/usr/bin/ld: cannot find -ludev
collect2: error: ld returned 1 exit status
Makefile:27: recipe for target 'dmrconfig' failed
make: *** [dmrconfig] Error 1

3rd step -- ~/md380/dmrconfig $ ls -l /usr/lib/x86_64-linux-gnu/libusb-1.0.a
-rw-r--r-- 1 root root 185176 Jun 24  2017 /usr/lib/x86_64-linux-gnu/libusb-1.0.a (so this looks correct per your output)

4th step -- ~/md380/dmrconfig $ dpkg -L libusb-1.0-0-dev
/.
/usr
/usr/include
/usr/include/libusb-1.0
/usr/include/libusb-1.0/libusb.h
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libusb-1.0.a
/usr/lib/x86_64-linux-gnu/pkgconfig
/usr/lib/x86_64-linux-gnu/pkgconfig/libusb-1.0.pc
/usr/share
/usr/share/doc
/usr/share/doc/libusb-1.0-0-dev
/usr/share/doc/libusb-1.0-0-dev/copyright
/usr/lib/x86_64-linux-gnu/libusb-1.0.so
/usr/share/doc/libusb-1.0-0-dev/README
/usr/share/doc/libusb-1.0-0-dev/changelog.Debian.gz
(Although you did not post your full output, I did in hopes you might spot something?)

5th step -- editing MAKEFILE.
Now the error (after removing the suggested from LIB)
Linux
ifeq ($(UNAME),Linux)
    OBJS        += hid-libusb.o
    LIBS        = -lpthread -ludev # per Sergev, removed Wl,-Bstatic -lusb-1.0 -Wl,-Bdynamic B4 compiling
endif

After running make, output here is --
~/md380/dmrconfig $ make
Makefile:1: *** empty variable name.  Stop.
So, I'm stumped and doing a bit of googling, am getting all sorts of suggestions, which I'm wary of trying.
Dennis KB7ST

**Comment by sergev** [comment]
Looks like you have some progress, so now it's a different error:
```
/usr/bin/ld: cannot find -ludev
```
Try this:
```
sudo apt install libudev-dev
```

**Comment by OH1KH** [comment]
Hi!
I had same problem with Fedora28.

Open Makefile and change "-llibusb-1.0" to the end of line here:
# Linux
ifeq ($(UNAME),Linux)
    OBJS        += hid-libusb.o
    LIBS        = -Wl,-Bstatic  -Wl,-Bdynamic -lpthread -ludev -lusb-1.0
endif

I do not know why, but for me it helped and dmrconfig was created and starts.

I have another problem: It cannot found radio:

[saku@hamtpad GD77]$ dmrconfig -r -t
Cannot find USB device 0483:df11
Cannot find USB device 15a2:0073
No radio detected.
Check your USB cable!
[saku@hamtpad GD77]$ lsusb
Bus 002 Device 004: ID 05c6:9204 Qualcomm, Inc. 
Bus 002 Device 007: ID 15a2:0073 Freescale Semiconductor, Inc.

Even it is there (poor USB cable, perhaps?)

**Comment by OH1KH** [comment]
Hi again!

Did not read all answers. Moving -llibusb-1.0 to the end does the same as sergev suggest ( just remove "-Wl,-Bstatic" and "-Wl,-Bdynamic" flags) .

And my problem was solved running program with sudo. Norma user does not have rights to access that usb device. Prehaps need to do mode=666 to udev rules (imagined that I can pass it, but no... ;-)

So in my case all seems to be solved. Sorry for QRM. Too fast to use, too little to read instructions.

**Comment by sergev** [comment]
OK, finally it works for you. The problem of permissions can be solved: see README.md.
As for libusb, I'll modify Makefile to link it statically when possible, and dynamically otherwise.

**Comment by sergev** [comment]
Solved in a4c5f893d2749727493427320c7f01768966ba51.

**Comment by sergev** [comment]
Great!
BTW, no need to always build from sources: binary packages are also available for download.
Latest version: [dmrconfig-1.1.zip](https://raw.githubusercontent.com/wiki/sergev/dmrconfig/releases/dmrconfig-1.1.zip)

For details, see: [dmrconfig/wiki](https://github.com/sergev/dmrconfig/wiki)
